### PR TITLE
Include button elements in DOM XSS Scan Rule

### PR DESCRIPTION
**Pull request by mrod23**
The goal of this MR is to have the DomXSS Plugin click on additional buttons within the DOM to see if XSS can be triggered.

The script is hitting buttons like this: `<button onclick="onCancelClicked()">Click me</button>` , but doesn't test buttons formatted like this: but not like this `<button type="button">Cancel</button>`

Tested against an internal site, screenshot here: 
image

### CODE REVIEW

**Comment by mrod23** [comment]
The goal of this MR is to have the DomXSS Plugin click on additional buttons within the DOM to see if XSS can be triggered. 

The script is hitting buttons like this: 
<button onclick="onCancelClicked()">Click me</button>

but not like this 
<button type="button">Cancel</button>

File: addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java
Comment (psiinon) [inline addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java:461]: You can combine this into just one line:
`inputElements.addAll(findHelper(driver, By.tagName("button")))`

**Comment by psiinon** [comment]
I like the principal here - theres definiately more chance of hitting DOM XSSs if we click on the buttons.
Can we tell they are buttons and not sendKeys to them? 
Strictly speaking they will no longer all be input elements so maybe rename that var and if possible test that they are input elements before calling sendKeys(..).
Can you explain how it avoiding to click on cancel buttons?

File: addOns/domxss/CHANGELOG.md
Comment (thc202) [inline addOns/domxss/CHANGELOG.md:20]: This should be added to the above "Changed" section.

**Comment by thc202** [comment]
Worth adding tests.

File: addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java
Comment (mrod23) [reply addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java:461]: Will do. I'll combine

File: addOns/domxss/CHANGELOG.md
Comment (mrod23) [reply addOns/domxss/CHANGELOG.md:20]: > Can you explain how it avoiding to click on cancel buttons?
> Can you explain how it avoiding to click on cancel buttons?

The main reason I wanted to add this is so the plugin will click the cancel buttons. We have JavaScript similar to the below on one of our sites. When the Cancel button is clicked, the JavaScript that's in the retUrl attribute is triggered on the user's page. 
```
            $("#DIAGLOG").dialog("option", "buttons", {
                "Cancel": function () {
                    window.location = '@retUrl';
                },
                "Submit": function () {
                    xxxxxxxx
                    });
                }
            });
```

Comment (mrod23) [reply addOns/domxss/CHANGELOG.md:20]: > Can we tell they are buttons and not sendKeys to them?
What's an example of a sendKey?

**Comment by mrod23** [comment]
@thc202 
> Worth adding tests.

What kind of tests are you thinking?

**Comment by thc202** [comment]
Tests that verify the buttons are actually clicked, with a case where it leads to XSS and another that doesn't.

**Comment by mrod23** [comment]
@psiinon
>    Can you explain how it avoiding to click on cancel buttons?

The main reason I wanted to add this is so the plugin will click the cancel buttons. We have JavaScript similar to the below on one of our sites. When the Cancel button is clicked, the JavaScript that's in the retUrl attribute is triggered on the user's page.

```
            $("#DIAGLOG").dialog("option", "buttons", {
                "Cancel": function () {
                    window.location = '@retUrl';
                },
                "Submit": function () {
                    xxxxxxxx
                    });
                }
            });
```

>    Can we tell they are buttons and not sendKeys to them?

What's an example of a sendKey?

**Comment by psiinon** [comment]
https://github.com/zaproxy/zap-extensions/blob/41d43cb9eb93bff3a4fe2dd58c304f49dbb2b779/addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java#L482 :)
sendKeys( ) is the method that submits the input to input fields, which obviously doesnt apply to buttons.
And +1 for tests - ones which find vulns. We have a reasonable set of tests for this add-on but just shout if you need some help.

**Comment by mrod23** [comment]
Ah the sendKeys() method. Sounds good I can put in a check . 

I may need help with the unit tests. Not seeing a test page I can use here: **http://public-firing-range.appspot.com**

**Comment by psiinon** [comment]
No, I imported all of the DOM XSS vulns that still worked from that site :D
They are all here: https://github.com/zaproxy/zap-extensions/tree/master/addOns/domxss/src/test/resources/org/zaproxy/zap/extension/domxss
If you can create some single pages like those which demo the vulns which require button input then we can help you create tests for those.

**Comment by mrod23** [comment]
@psiinon 

I created a test page and added to the Resources directory. Zap will only find the XSS with the changes added in this MR. The current version shouldn't be able to find them. 

 In order for Zap to trigger an alert on the test page it requires a URL parameter called "returnUrl". 
Something like this: https://localhost:8080/test/jquery-3.html?returnUrl=test

I couldn't figure out how to get the Unit test to query a page with a URL parameter. Can you help me out with that? I started a unit test, but I have it commented out. 

I addressed the other comments mentioned above in the latest PR.

File: addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java
Comment (mrod23) [reply addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java]: combined

**Comment by kingthorin** [comment]
@mrod23 have a look at https://github.com/zaproxy/zap-extensions/blob/master/addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java it should have details to help you out.

**Comment by mrod23** [comment]
Thanks @kingthorin 

@kingthorin  / @psiinon  can you take a look at my branch when you get a chance?
The unit test is still failing, pasted what the failure output looks like 

I think I need to update when the code raises an assertAlert here: https://github.com/zaproxy/zap-extensions/blob/ff51107d3abf016a51845b065ac3523abe9a1e5c/addOns/domxss/src/test/java/org/zaproxy/zap/extension/domxss/DomXssScanRuleUnitTest.java#L127

Tried a few different combinations of new variables but I couldn't get it to trigger. 

If you serve up the CancelButton page locally, and run Zap with the plugin built from this branch it'll trigger an alert. Pasting screenshot below 


image


image

**Comment by kingthorin** [comment]
Will try to look this evening.

**Comment by thc202** [comment]
One of problems is that the proxy created by the scan rule does not have the Root CA cert thus failing to access the jQuery scripts. That can be addressed by adding to the test `setup` method `CachedSslCertifificateServiceImpl.getService().initializeRootCA(SslCertificateUtils.createRootCA());` but I'd suggest adding the required resources locally to not rely on the remote ones.

**Comment by mrod23** [comment]
Winded up adding the JQuery scripts directly to the CancelButton.html page. 

This is now passing the build. 

Ready for review, please

**Comment by mrod23** [comment]
@psiinon  / @thc202 
Can you please take a look at this MR when you get a chance? Think it's ready for approval.

File: addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java
Comment (psiinon) [inline addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java:487]: I agree these names are more descriptive, but what if people are already using the existing ones?
No idea if thats the case, but if so then it would break things for them...

**Comment by psiinon** [comment]
Woah, the CancelButton.html file is huge :P
Including a copy of jquery makes it harder to maintain.
I'd much rather a simple html file with minimal vanilla JS as per the other tests. The vast majority of the JS in that file is not needed...

**Comment by psiinon** [comment]
Other than those comments - its getting very close :)

**Comment by thc202** [comment]
Would you mind squash the commits?

**Comment by psiinon** [comment]
As an aside, clicking on one button could change the state in a way that impacts another button.
For example, there could be a form with a 'Clear' button which just cleas the fields.
The rule could hit the clear button, clearing the fields, before hittin another button which would trigger the XSS.
Dont think that needs to be addressed in this PR as it could get v complicated, just worth thinking about for the future. If we have a simple native JS example which the rules works with the we would hopefully be able to tweak that to create a false negative example that we'd want to fix at some point...

**Comment by kingthorin** [comment]
https://github.com/zaproxy/zap-extensions/pull/2638/checks?check_run_id=1524550292

To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

File: addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java
Comment (mrod23) [reply addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java]: Updated to possibleDomXSSTriggers2

**Comment by mrod23** [comment]
@psiinon  updated, pulled out the JQuery and using my own JavaScript. 
@thc202  squashed commits

**Comment by mrod23** [comment]
@psiinon  / @thc202 
I think this PR is ready to go now. Please review/approve when you get a chance

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thanks!

**Review by kingthorin — APPROVED** [review]
Thanks

**Pull request merged**
